File: src/j/implementation/microSet/InnerHashSet.java
package j.implementation.microSet;

import java.util.HashSet;
import java.util.Iterator;

public class InnerHashSet<T> extends HashSet<T> implements InnerSet<T>
{

	public InnerHashSet() {
        super();
    }

    public InnerHashSet(InnerSet i) {
        this();
        for (InnerSetIterator<T> iterator = new InnerSetIterator<>(i); iterator.hasNext(); ) {
            add(iterator.next());
        }
    }

    public InnerHashSet(InnerSet<T> i, T p_element) {
        this();
        for (InnerSetIterator<T> iterator = new InnerSetIterator<>(i); iterator.hasNext(); ) {
            add(iterator.next());
        }
        add(p_element);
    }

    @Override
    public InnerSet<T> addElement(T p_element) {
        add(p_element);
        return this;
    }

    @Override
    public InnerSet<T> addUnChecked(T p_element) {
        return addElement(p_element);
    }

    @Override
    public InnerSet<T> addAllElements(InnerSet<T> innerSet) {
        for (T el : this) {
            innerSet = innerSet.addElement(el);
        }
        return innerSet;
    }

    @Override
    public boolean containsElement(Object p_element) {
        return contains(p_element);
    }

    @Override
    public boolean containsAllElements(InnerSet<T> innerSet) {
        //return containsAll(innerSet);
        return false;
    }

    @Override
    public T getElement(int index) {
        return null;
    }

    @Override
    public InnerSet<T> copy() {
        return new InnerHashSet<>(this);
    }

    @Override
    public InnerSet<T> removeElement(Object p_element) {
        remove(p_element);
        return this;
    }

    @Override
    public InnerSet<T> removeAllElements(InnerSet<T> innerSet) {
        //removeAll(innerSet);
        return null;
    }

    @Override
    public InnerSet<T> retainAllElements(InnerSet<T> innerSet) {
       // retainAll(innerSet);
        return this;
    }

    @Override
    public InnerSet<T> addAndPropagate(T p_element, MicroSet<T> microSetToPropagate) {
        if (!contains(p_element)) {
            microSetToPropagate.add(p_element);
            add(p_element);
        }
        return this;
    }

    public InnerSet<T> addAllAndPropagate(InnerSet<T> innerSet, MicroSet<T> microSetToPropagate) {
        return innerSet.addAllAndPropagateReverse(this, microSetToPropagate);
    }

    @Override
    public InnerSet<T> addAllAndPropagateReverse(InnerSet<T> innerSet, MicroSet<T> microSetPropagate) {
        for (T el : this) {
            innerSet = innerSet.addAndPropagate(el, microSetPropagate);
        }
        return innerSet;
    }

    @Override
    public int getSize() {
        return size();
    }

    @Override
    public InnerSet<T> clear(boolean unused) {
        this.clear();
        return this;
    }
}

File: src/scala/implementation/InnerHashSet.scala
package scala.implementation

import java.util.HashSet



 class InnerHashSet[T] extends HashSet[T] with InnerSet[T]{

  def this(p_innerSet: InnerSet[T]) =
  {
    this
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext)
    {
      add(it.next())
    }
  }

  def this(p_innerSet: InnerSet[T], p_element: T) =
  {
    this
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext)
    {
      add(it.next())
    }
    add(p_element)

  }

  override def addElement(p_element: T): InnerSet[T] =
  {
    if(!contains(p_element)) add(p_element)
    this
  }

  override def addUnchecked(p_element: T): InnerSet[T] =
  {
    add(p_element)
    this
  }

   override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
   {
     if(!contains(p_element))
     {
       p_microSet.add(p_element)
       add(p_element)
     }
     this
   }

  override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
  {
    val it: InnerSetIterator[T] = this.iterator
    var innerSetReturn: InnerSet[T] = p_innerSet
      while(it.hasNext){
      innerSetReturn = innerSetReturn.addElement(it.next())
    }
    innerSetReturn
  }

   override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] = p_innerSet.addAllAndPropagateReverse(this, p_microSet)

   override def addAllAndPropagateReverse(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =
   {
     val it: InnerSetIterator[T] = this.iterator
     var innerSetReturn: InnerSet[T] = p_innerSet
       while(it.hasNext){
       innerSetReturn = innerSetReturn.addAndPropagate(it.next(),p_microSet)
     }
     innerSetReturn
   }

   override def containsElement(p_element: Object): Boolean = contains(p_element)

  override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =
  {
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext){
      if(!contains(it.next())) return false
    }
    true
  }

  override def getElement(index: Int): Option[T] = ??? //TODO

  override def removeElement(p_element: Object): InnerSet[T] =
  {
    remove(p_element)
    this
  }

  override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
  {
    val it: InnerSetIterator[T] = p_innerSet.iterator
    while(it.hasNext){
      if(contains(it.next())) remove(it.next())
    }
    this
  }

  override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] = ??? //TODO

  override def getSize: Int = ??? //TODO

   override def iterator: InnerSetIterator[T] = new InnerSetIterator(this)

  override def copy: Option[InnerHashSet[T]] =
  {
    Some(new InnerHashSet(this))
  }

  override def clear(unused: Boolean): InnerSet[T] = this.clear(unused)


 }
